webapi/src/main/scala/org/knora/webapi/messages/admin/responder/projectsmessages/ProjectsMessagesADM.scala

```diff
@@ -505,130 +511,104 @@ case class ProjectADM(
-
-    new ProjectIdentifierADM(
-      maybeIri =
-        sf.validateAndEscapeOptionalProjectIri(maybeIri, throw BadRequestException(s"Invalid user project $maybeIri")),
-      maybeShortname = sf.validateAndEscapeOptionalProjectShortname(
-        maybeShortname,
-        throw BadRequestException(s"Invalid user project shortname $maybeShortname")
-      ),
-      maybeShortcode = sf.validateAndEscapeOptionalProjectShortcode(
-        maybeShortcode,
-        throw BadRequestException(s"Invalid user project shortcode $maybeShortcode")
-      )
-    )
-  }
-}
+sealed trait ProjectIdentifierADM { self =>
+  def asIriOption: Option[String] =
+    self match {
+      case IriIdentifier(value) => Some(value.value)
+      case _                    => None
+    }
 
-/**
- * Represents the project's identifier. It can be an IRI, shortcode or shortname.
- *
- * @param maybeIri       the project's IRI.
- * @param maybeShortname the project's shortname.
- * @param maybeShortcode the project's shortcode
- */
-class ProjectIdentifierADM private (
-  maybeIri: Option[IRI] = None,
-  maybeShortname: Option[String] = None,
-  maybeShortcode: Option[String] = None
-) {
-
-  // squash and return value.
-  val value: String = List(
-    maybeIri,
-    maybeShortname,
-    maybeShortcode
-  ).flatten.head
-
-  def hasType: ProjectIdentifierType =
-    if (maybeIri.isDefined) {
-      ProjectIdentifierType.IRI
-    } else if (maybeShortcode.isDefined) {
-      ProjectIdentifierType.SHORTCODE
-    } else {
-      ProjectIdentifierType.SHORTNAME
+  def asShortcodeOption: Option[String] =
+    self match {
+      case ShortcodeIdentifier(value) => Some(value.value)
+      case _                          => None
     }
 
-  /**
-   * Tries to return the value as an IRI.
-   */
-  def toIri: IRI =
-    maybeIri.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.IRI' type.")
-    )
+  def asShortnameOption: Option[String] =
+    self match {
+      case ShortnameIdentifier(value) => Some(value.value)
+      case _                          => None
+    }
 
-  /**
-   * Returns an optional value of the identifier.
-   */
-  def toIriOption: Option[IRI] =
-    maybeIri
+  def asUuidOption: Option[String] =
+    self match {
+      case UuidIdentifier(value) => Some(UuidIdentifier.makeProjectIri(value.value))
+      case _                     => None
+    }
+}
+
+object ProjectIdentifierADM {
 
   /**
-   * Tries to return the value as an SHORTNAME.
+   * Represents [[IriIdentifier]] identifier.
+   *
+   * @param value that constructs the identifier in the type of [[ProjectIri]] value object.
    */
-  def toShortname: String =
-    maybeShortname.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.SHORTNAME' type.")
-    )
+  final case class IriIdentifier(value: ProjectIri) extends ProjectIdentifierADM
+  object IriIdentifier {
```

Comment:
I would call it `ProjectIriIdentifier` in order to not confuse it with other types of IRIs.

Reply:
It lives in `ProjectIdentifierADM` so isn't it enough to know that it's about `ProjectIri`? I like informative naming but we should decide if we go for last tail having all information, or make it a bit shorter if object where it lives already have it, so:
`ProjectIdentifierADM. IriIdentifier` vs `ProjectIdentifierADM. ProjectIriIdentifier`

Reply:
Especially there are no more `IriIdentifier`, the only one exists for the Project

Reply:
If it is used together with the object, as `ProjectIdentifierADM.IriIdentifier`, then it is enough.